### DISCUSSION TITLE
Cannot create Program from binary

### DISCUSSION THREAD

**Discussion by yves-surrel** [pyopencl/pyopencl#544] [Troubleshooting]
From the [doc](https://documen.tician.de/pyopencl/runtime_program.html) `class pyopencl.Program(context, devices, binaries)`, existing binaries can be provided to create a `Program`. However, this simple example fails:

```
import os
import pyopencl as cl

# Kernel function
src = """
__kernel void test(
    __global int *a)
{
    a[get_global_id(0)] = 0;
}
"""

platforms = cl.get_platforms()

for platform in platforms:
    devices = platform.get_devices()
    for i_device, device in enumerate(devices):
        ctx = cl.Context([device])
        prg = cl.Program(ctx, src).build()
        binary, = prg.get_info(cl.program_info.BINARIES)
        # Create Program from binary
        prg1 = cl.Program(ctx, (device,), (binary,))
        try:
            print(prg1.all_kernels())
        except cl.LogicError as e:
            print(e)
```
The result is:
```
clCreateKernelsInProgram failed: INVALID_PROGRAM_EXECUTABLE
```
for all devices found in the platform.

MacOS 11.6.4 Big Sur, Macbook Pro.
On the Windows side (BootCamp), it crashes Python.

**Comment by inducer** [comment]
You need:
```
prg1 = cl.Program(ctx, (device,), (binary,)).build()
```
(Yes, even binaries need to be built. Look it up in the spec. :)

**Comment by yves-surrel** [comment]
So simple! Actually, I had guessed that long ago. Unfortunately, it does not work:
```
import os
import pyopencl as cl

# Kernel function
src = """
__kernel void test(
    __global int *a)
{
    a[get_global_id(0)] = 0;
}
"""

platforms = cl.get_platforms()

for platform in platforms:
    devices = platform.get_devices()
    for i_device, device in enumerate(devices):
        ctx = cl.Context([device])
        prg = cl.Program(ctx, src).build()
        binary, = prg.get_info(cl.program_info.BINARIES)
        # Create Program from binary
        try:
            prg1 = cl.Program(ctx, (device,), (binary,)).build()
        except cl.RuntimeError as e:
            print(1, e)
```
prints:
```
1 clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE

Build on <pyopencl.Device 'Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz' on 'Apple' at 0xffffffff>:


(options: -I /Users/yves/miniconda3/lib/python3.8/site-packages/pyopencl-2021.2.6-py3.8-macosx-10.9-x86_64.egg/pyopencl/cl)
```
You can see two things:
1 - Still an error
2 - The exception is not properly catched (there should be three lines `1` because I have three devices.

EDITED : 
BTW, I have re-re-read the doc, I see nowhere that program created from built binaries should be built (again)...

**Reply by inducer** [reply]
> I have re-re-read the doc, I see nowhere that program created from built binaries should be built (again)...

The CL 3.0 Spec, [Section 5.8.4](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_building_program_executables) says:

> [clBuildProgram](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clBuildProgram) must be called for program created using [clCreateProgramWithSource](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clCreateProgramWithSource), [clCreateProgramWithIL](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clCreateProgramWithIL) or [clCreateProgramWithBinary](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clCreateProgramWithBinary) to build the program executable for one or more devices associated with program.

> The exception is not properly catched (there should be three lines 1 because I have three devices.

Why not catch https://documen.tician.de/pyopencl/runtime.html#pyopencl.Error instead of `cl.RuntimeError`?

**Comment by yves-surrel** [comment]
> The CL 3.0 Spec

Ah ah! I was concentrated on the pyOpenCL doc. I humbly suggest that a word about this is added in the pyOpenCL `class pyopencl.Program(context, devices, binaries)` documentation about this...

> Why not catch https://documen.tician.de/pyopencl/runtime.html#pyopencl.Error instead of cl.RuntimeError?

Because without the `try...except`structure, the output is:
```
RuntimeError: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE

Build on <pyopencl.Device 'Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz' on 'Apple' at 0xffffffff>:
```
However, using `except cl.Error as e:`or even the generic `except Exception as e: ` does not change anything.

### Latest news
Only the device _Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz_ raises the Exception. The code works for all other 2 devices on my MBP. Does it provide you any hint?

BTW this explain why there is only one exception catched! Sorry...

FYI the binary bytes content is
```
bplist00[d4][01][02][03][04][05][06][07][08]_[10][0f]clBinaryVersion\clBinaryData_[10][11]clPlatformVersion^clBinaryDriver[11][01][02]O[10]5/private/var/db/CVMS/cvmsCodeSignObjIs1zWPcqldHAlFQQ[00]_[10]!OpenCL 1.2 (Sep  5 2021 22:39:07)S1.1[08][11]#0DSV[8e][b2][00][00][00][00][00][00][01][01][00][00][00][00][00][00][00][09][00][00][00][00][00][00][00][00][00][00][00][00][00][00][00][b6]
```
where non printable bytes are replaced by [xx]. Seems to be an intermediate representation (IR) as per Khronos OpenCL doc, but it should work the same.

**Reply by inducer** [reply]
> I humbly suggest that a word about this is added in the pyOpenCL `class pyopencl.Program(context, devices, binaries)` documentation about this...

I mostly don't want to do this, otherwise I'd be in the business of ultimately copy-pasting the entire spec into the pyopencl docs, which would lead nowhere good.

**Comment by yves-surrel** [comment]
I understand...

**Comment by yves-surrel** [comment]
Any idea why _Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz_ is not building ? Am I missing some option (e.g. about caching)? AFAIU, pyopencl is internally using disk caching so...?
